### PR TITLE
Fix bug in config_parse.py when batch_norm layer is used in RecurrentLayerGroup

### LINKED ISSUE #961: batch_norm layer 放在 RecurrentLayerGroup 里面时，配置解析的一个bug
batch norm layer 作为 RecurrentLayerGroup step 函数里面的一个layer，配置解析时会出错。

比如下面的layer A 在recurrent layer group 里面， layer_group 的name 是 ”layer_group“：
```
    Layer(name="B",
          type='batch_norm',
          active_type="relu",
          inputs=Input("A", image=Image(channels=128, img_size=1)), )
```
调用  inputs=Input("A", image=Image(channels=128, img_size=1)), ) 时，input[0].input_layer_name 会被 MakeLayerNameInParentSubmodel  函数命名成 ：A@layer_group。

然后batch norm layer 在解析配置时会继续调用下面的代码（config_parser.py 的 1840 ~ 1847 行）隐式地添加 mean 和 std ：

```
        for i in xrange(2):
            inputs.append(
                Input(
                    inputs[0].input_layer_name,
                    initial_std=0.0,
                    initial_mean=0.0,
                    is_static=True,
                    is_shared=is_shared, ))
```
这里的 Input 会再次调用 MakeLayerNameInParentSubmodel ， 将 inputs[0].input_layer_name （这时已经是  A@layer_group）命名为 ：A@layer_group@layer_group，这一次重命名会引起错误。

### PR DESCRIPTION
**Pull request by qingqing01**
Fix #961 

### 解决方法：

在 config_parse.py里构造 BatchNormLayer的另外两个Input()时，不经过函数 https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer/config_parser.py#L503 

### 错误原因：

产生原因是 BatchNorm Layer里自动为moving average参数添加了Input()：
```
        for i in xrange(2):
            inputs.append(
                Input(
                    inputs[0].input_layer_name,
                    initial_std=0.0,
                    initial_mean=0.0,
                    is_static=True,
                    is_shared=is_shared,  ))
```
因为调用顺序原因，这里引用的名字 `inputs[0].input_layer_name`,   在RecurrentLayerGroup里是被函数MakeLayerNameInSubmodel()编码之后的名字，即  ``` input_layer_name + "@" +  submodel_name ```， 导致出错。

##  其他解决办法：

如果在config_parse.py外面， 由用户自定添加，例如下面方式，这样两个Input(1)和Input(2)的name经过config_parse.py内部编码会与 Input(0) 保持一致， 不会出错。

```
    Layer(name=name + 'batch_norm1',
          type='batch_norm',
          batch_norm_type=bn_type,
          active_type="relu",
          use_global_stats=use_global_stats,
          bias=Bias(initial_mean=0.0,
                    initial_std=0.0,
                    is_static=static,
                    learning_rate=lr),
          inputs=[Input(name + 'conv1',
                                initial_mean=0,
                                initial_std=0.1,
                                is_static=static,
                                learning_rate=lr,
                                image=Image(channels=size,
                                        img_size=1)),
                       Input(name + 'conv1',
                                initial_std=0.0,
                                initial_mean=0.0,
                                is_static=True, ),
                       Input(name + 'conv1',
                                initial_std=0.0,
                                initial_mean=0.0,
                                is_static=True, ),  ] )
```

这个可以在 `trainer_config_helpers` 里重新包装， 但是部分还是用老配置的用户，得手动自己添加，较麻烦。

其他办法：

   -  是否可以在config_parse.py里只添加 Parameter()， 而不添加Input()？
      但Layer.cpp里init函数https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/Layer.cpp#L57  解析时一个input对应一个参数，所以应该不可行。

   -  可否合并为同一个参数？ 
     因为 .w0 .w1, .w2的一些操作不同，也无法创建同一片空间，在Layer.cpp里用offset的方式构造。

### CODE REVIEW

File: python/paddle/trainer/config_parser.py [hunk 1]
```diff
@@ -498,9 +498,12 @@ def __init__(
             is_static=None,
             is_shared=None,
             update_hooks=None,
-            input_layer_argument=None, ):
+            input_layer_argument=None,
+            not_make_layer_name_in_submodel=None, ):
```
Comment (reyoung) [inline python/paddle/trainer/config_parser.py]: 请问这个变量到底之bool型的还是一个指针呀？

python经常把默认参数设置成None是有原因的，参考 http://effbot.org/zone/default-values.htm

对于简单类型，比如整数，bool型，那么默认值直接传就可以了。因为Python对于简单类型是值传递而不是引用传递。
